src/Backups/BackupCoordinationReplicatedSQLObjects.h

```diff
@@ -0,0 +1,49 @@
+#pragma once
+
+#include <Core/Types.h>
+#include <map>
+#include <unordered_set>
+
+
+namespace DB
+{
+enum class UserDefinedSQLObjectType;
+
+/// This class is used by hosts to coordinate the user-defined SQL objects they're going to write to a backup.
+/// It's designed to make all hosts save the same objects to the backup even in case some objects change while
+/// the backup is being produced. This is important to make RESTORE more predicitible.
+///
+/// For example, let's consider three replicas having access to the user-defined function `f1`.
+/// This class ensures that the following files in the backup will be the same:
+/// /shards/1/replicas/1/data/system/functions/f1.sql
+/// /shards/1/replicas/2/data/system/functions/f1.sql
+/// /shards/1/replicas/3/data/system/functions/f1.sql
+///
+/// To implement that this class chooses one host to write user-defined SQL objects for all the hosts so in fact all those files
+/// in the example above are written by single host.
+
+class BackupCoordinationReplicatedSQLObjects
+{
+public:
+    BackupCoordinationReplicatedSQLObjects();
+    ~BackupCoordinationReplicatedSQLObjects();
+
+    /// Adds a path to directory keeping user defined SQL objects.
+    void addDirectory(const String & loader_zk_path, UserDefinedSQLObjectType object_type, const String & host_id, const String & dir_path);
+
+    /// Returns all added paths to directories if `host_id` is a host chosen to store user-defined SQL objects.
+    Strings getDirectories(const String & loader_zk_path, UserDefinedSQLObjectType object_type, const String & host_id) const;
+
+private:
+    using ZkPathAndObjectType = std::pair<String, UserDefinedSQLObjectType>;
```

Comment:
As for 1:
1) First, I want to point that it is consistent with [code](https://github.com/ClickHouse/ClickHouse/blob/5b423837306831277577cbb9a9aca9031f06421f/src/Backups/BackupCoordinationReplicatedAccess.h#L38) for access_storage.
2) The `...Type` here is a part of `...ObjectType` and not the whole alias. It denotes second element of pair. Just `Object` would be incorrect.
3) `std::pair` is not always error-prone. For example, like here it is used as a key for `map` which is quite a common practice. Because it has predefined comparisons operators (<=> actually). And it is not accessed via it `first, second` which is really ugly.

As for 2:
1) Frankly, i don't see need to introduce another indirection level for parameters. At least for now.